src/communication/communicator.hpp

```diff
@@ -45,31 +48,71 @@ class communicator {
+        using util::make_span;
+        num_domains_ = comms_.size();
+        num_local_groups_ = dom_dec.num_local_groups();
+
+        // For caching information about each cell
+        struct gid_info {
+            using connection_list = decltype(rec.connections_on(0));
+            cell_gid_type gid;
+            cell_gid_type local_group;
+            connection_list conns;
+            gid_info(cell_gid_type g, cell_gid_type lg, connection_list c):
+                gid(g), local_group(lg), conns(std::move(c)) {}
+        };
+
+        // Make a list of local gid with their group index and connections
+        //   -> gid_infos
+        // Count the number of local connections (i.e. connections terminating on this domain)
+        //  -> n_cons: scalar
+        // Calculate and store domain id of the presynaptic cell on each local connection
+        //  -> src_domains: array with one entry for every local connection
+        // Also the count of presynaptic sources from each domain
+        //  -> src_counts: array with one entry for each domain
+        std::vector<gid_info> gid_infos;
+        gid_infos.reserve(dom_dec.num_local_cells());
+
+        cell_local_size_type n_cons = 0;
+        std::vector<unsigned> src_domains;
+        std::vector<cell_size_type> src_counts(num_domains_);
+        for (auto i: make_span(0, num_local_groups_)) {
+            const auto& group = dom_dec.get_group(i);
+            for (auto gid: group.gids) {
+                gid_info info(gid, i, rec.connections_on(gid));
+                n_cons += info.conns.size();
+                for (auto con: info.conns) {
+                    const auto src = dom_dec.gid_domain(con.source.gid);
+                    src_domains.push_back(src);
+                    src_counts[src]++;
+                }
+                gid_infos.push_back(std::move(info));
+            }
+        }
 
-    /// builds the optimized data structure
-    /// must be called after all connections have been added
-    void construct() {
-        if (!std::is_sorted(connections_.begin(), connections_.end())) {
-            threading::sort(connections_);
+        // Construct the connections.
+        // The loop above gave the information required to construct in place
+        // the connections as partitioned by the domain of their source gid.
+        connections_.resize(n_cons);
+        connection_part_ = algorithms::make_index(src_counts);
+        auto offsets = connection_part_;
+        std::size_t pos = 0;
+        for (const auto& cell: gid_infos) {
+            for (auto c: cell.conns) {
+                const auto i = offsets[src_domains[pos]]++;
+                connections_[i] = {c.source, c.dest, c.weight, c.delay, cell.local_group};
+                ++pos;
+            }
         }
+
+        // Sort the connections for each domain.
+        // This is num_domains_ independent sorts, so it can be parallelized trivially.
+        const auto& cp = connection_part_;
```

Comment:
I think that we are safe here. The communicator is constructed before the cell_groups. I think that it should be possible to construct the communicator at the same time that the cell_groups are being constructed, because the work is independent (they both take the recipe and domain_decomposition as read only input parameters).